Ep/ Rename Spec to AffineScheme #3345

paemurru opened the pull request:
Resolves #3345

paemurru committed:
Rename Spec -> AffineScheme

paemurru committed:
Fix missed Spec -> AffineScheme

paemurru committed:
Fix: rename one more Spec to AffineScheme

paemurru committed:
Fix: change three more Spec's to AffineScheme

paemurru committed:
Rename the function AffineScheme to spec

Use `AffineScheme` only for the type, not the function.

paemurru committed:
Rename proj for toric divisors to projectivization

As was recommended by Martin Bies and Wolfram Decker.

paemurru committed:
Change non-scheme proj functions to something else

For biproducts, naming the output triple
    [obj], inj, pr
seems to be common. Among the changes was changing "proj" to "pr" in the
triples
    [obj], inj, proj

paemurru committed:
Function ProjectiveScheme to projective_scheme

Also some other minor fixes

paemurru committed:
Fix: spec now extends the constructor AffineScheme

In particular, spec allows the argument to be a quotient ring or localized ring. Similarly for proj. Probably the tests would not have passed in the previous commit because of this

paemurru committed:
rename filename

paemurru committed:
Fix tests for proj

paemurru committed:
projecitve_scheme -> proj

paemurru committed:
Add checks for being standard graded

paemurru committed:
Remove possibly redundant proj function

proj(Q::MPolyQuoRing{MPolyDecRingElem{T, PT}}) where {T, PT<:MPolyRingElem{T}} = ProjectiveScheme(Q)

paemurru committed:
Fix docs

paemurru committed:
Fix duplicate exports entry

paemurru committed:
Use script to rename all other functions and types

paemurru committed:
Fix sorting script, sort src/exports.jl

paemurru committed:
Rename two test files

joschmitt commented on experimental/QuadFormAndIsom/test/setup_tests.jl:
```diff
@@ -1 +0,0 @@
-../../../test/Serialization/setup_tests.jl
```
This file is actually a symbolic to `../../../test/Serialization/setup_tests.jl`. I assume it is a copy now because you used `sed`. (At least that's what happened to me in the past.)

paemurru replied:
Fixed. I don't know if there are other symbolic links ruined

paemurru committed:
Fix symbolic link

paemurru committed:
Strange biproduct order in hermitian_miranda_morrison.jl

Remove commented question that I made earlier. Better to make an issue or ask someone than comment the code.

paemurru committed:
Merge branch 'oscar-system:master' into EP/RenameSpecToAffineScheme

lgoettgens commented on experimental/LieAlgebras/src/LieAlgebraModuleHom.jl:
```diff
@@ -403,7 +403,7 @@ function canonical_projection(V::LieAlgebraModule, i::Int)
   @req 1 <= i <= length(Vs) "Index out of bound"
   j = sum(dim(Vs[l]) for l in 1:(i - 1); init=0)
-  proj = hom(
+  projection = hom(
```
What is the purpose of this change and the similar ones in parts of the code that have nothing to do with schemes? I don't really see why they are necessary.

paemurru replied:
https://github.com/oscar-system/Oscar.jl/issues/3345#issuecomment-1950227788

lgoettgens replied:
I am with @thofma here: https://github.com/oscar-system/Oscar.jl/issues/3345#issuecomment-1950237240

thofma reviewed (dismissed):
Variables can be called whatever the original author wants (unless of course it results now in a conflict). This should not be changed here.

lgoettgens commented:
I think this can be achieved by partially reverting https://github.com/oscar-system/Oscar.jl/pull/3425/commits/cb13aad4d1fd142c6da5d4274d4f27746e1adcdc

paemurru commented:
@thofma @lgoettgens. The changes in https://github.com/oscar-system/Oscar.jl/commit/cb13aad4d1fd142c6da5d4274d4f27746e1adcdc are mostly about how to name the maps returned by direct sums and products. In algebraic geometry, naming the returned projection map `proj` conflicts with the homogeneous spectrum. In areas that have connections to algebraic geometry, it is more subjective where to draw the line, where to keep `proj` for internal variable names and where not.

In general, I believe there is value in having consistent naming for variables returned in direct sums and products, so I mostly went with `inj` and `pr`, as that was already common.

OSCAR project leader Michael Joswig encouraged not using `proj` outside of the algebraic geometry meaning, see https://github.com/oscar-system/Oscar.jl/issues/3345#issuecomment-1950227788. Unless you can quickly convince one of the OSCAR project leaders why your point of view is better, I don't think this justifies blocking the merging, especially given our time constraints. If you wish to later change some of the internal variable names back to `proj` in a different pull request, you are welcome to, it is a trivial matter.

fieker commented:
I as an author name variables in my code. No one else. I as well as several others would be extremely pissed if anyone, for any reason but syntax errors, changes purely internal names. Sorry

thofma commented:
I think there was a misunderstanding in https://github.com/oscar-system/Oscar.jl/issues/3345#issuecomment-1950227788, which was about functions called `proj` in the global Oscar namespace (which is why there is a reference to "be[ing] globally consistent") and not local variable names. But there does not exist such a function, so there is nothing to adjust.

paemurru commented:
I think that by "globally consistent" they meant "everywhere consistent" in https://github.com/oscar-system/Oscar.jl/issues/3345#issuecomment-1950227788

paemurru commented:
If there is a particular change you wish to make regarding the renaming, please make it explicit and I am happy to make the change

lgoettgens commented:
> If there is a particular change you wish to make regarding the renaming, please make it explicit and I am happy to make the change

As I already wrote earlier, I think this is about everything in https://github.com/oscar-system/Oscar.jl/commit/cb13aad4d1fd142c6da5d4274d4f27746e1adcdc (but the very few changes regarding to `projectivization` in it). Please correct me if I am wrong @fieker @thofma.

paemurru commented:
Do I understand correctly that you wish to revert the changes to variables named `proj` in the sense of a projection in local scope in https://github.com/oscar-system/Oscar.jl/commit/cb13aad4d1fd142c6da5d4274d4f27746e1adcdc, regardless of whether they appear in group theory, algebra or algebraic geometry, regardless of whether they could be confused with the homogeneous spectrum?

fieker commented:
There is no reason that I can see why an internal variable change is warranted.
If there is a convincing reason, I'd like learn it

paemurru commented:
@fieker Easier to maintain code, find errors in code, collaborate with others. The same reason why style guides exist in most programming languages, why linters and automatic code formatters exist. The same reason why all code, including internal functions and variables in local scope, are usually formatted according to the style guide.

In this particular case, it is a common best practice not name a variable after functions or variables in global scope (see [StackOverflow](https://stackoverflow.com/a/25000042)). It is something that linters warn against (I haven't found any currently maintained Julia linter, but here is an example from [Pylint](https://pylint.pycqa.org/en/latest/user_guide/messages/warning/redefined-outer-name.html)).

paemurru commented:
Just to reiterate, I am happy to change the code as you suggest. I just want to make sure I understand what it is you wish to change, as it seems strange to me to use ´proj´ in algebraic geometry for projection morphisms where it can easily be confused with the homogeneous spectrum. It seems strange to me to go against standard best practices and to go against an OSCAR project leader's recommendation.

This renaming issue has been going on for weeks, with different people having different points of view, going back and forth on what or how to rename and with me having to accommodate to these new decisions. I am happy to make the changes, I just hope that I don't need to undo these changes again.

thofma commented:
> As I already wrote earlier, I think this is about everything in [cb13aad](https://github.com/oscar-system/Oscar.jl/commit/cb13aad4d1fd142c6da5d4274d4f27746e1adcdc) (but the very few changes regarding to `projectivization` in it).

paemurru committed:
Merge branch 'oscar-system:master' into EP/RenameSpecToAffineScheme

paemurru committed:
Undo everything in cb13aad except projectivization

paemurru committed:
Fix one forgotten proj from last commit

paemurru commented:
Undid all the changes in https://github.com/oscar-system/Oscar.jl/commit/cb13aad4d1fd142c6da5d4274d4f27746e1adcdc except for `proj` -> `projectivization`, as requested by @thofma. This includes changes in files not relating to schemes as well as changes in `Schemes`, changes in documentation and changes that had been approved by the original author in private communication.

thofma commented:
Thanks! Maybe @simonbrandhorst could give a quick glance and approve the changes here?

fingolfin commented:
@paemurru since you keep referring to "an OSCAR project leader's recommendation", let me point out that @fieker and @thofma also are OSCAR project leaders and as such I find it odd to draw that card against them... You can of course request that "the OSCAR leaders" sit together and give a final verdict before you change anything, so that you don't end up changing things back and forth (a quite reasonable request).

simonbrandhorst commented on src/AlgebraicGeometry/Schemes/ProjectiveSchemes/Objects/Attributes.jl:
```diff
@@ -485,7 +485,7 @@ homogeneous_coordinate_ring(P::ProjectiveScheme) = P.S
 projective_scheme_type(A::T) where {T<:AbstractAlgebra.Ring} = projective_scheme_type(typeof(A))
 projective_scheme_type(::Type{T}) where {T<:AbstractAlgebra.Ring} =
-ProjectiveScheme{T, mpoly_dec_ring_type(mpoly_ring_type(T))}
+proj{T, mpoly_dec_ring_type(mpoly_ring_type(T))}
```
This is a type getter and should therefore return the type.
```suggestion
ProjectiveScheme{T, mpoly_dec_ring_type(mpoly_ring_type(T))}
```

simonbrandhorst commented on src/AlgebraicGeometry/Schemes/ProjectiveSchemes/Objects/Constructors.jl:
```diff
@@ -2,13 +2,26 @@
-projective_scheme(S::MPolyDecRing) = ProjectiveScheme(S)
+function proj(S::MPolyDecRing)
+  is_standard_graded(S) || error("ring must be standard graded")
+  return ProjectiveScheme(S)
```
It seems to me that the better place for this input check would be the actual constructor
`ProjectiveScheme`? (however this need not delay merging this pull request.)

simonbrandhorst commented on src/Groups/directproducts.jl:
```diff
@@ -136,7 +136,7 @@ end
 
 The keyword argument `morphisms` is `false` by default. If it is set `true`, then
-the output is a triple (`G`, `emb`, `proj`), where `emb` and `proj` are the
+the output is a triple (`G`, `emb`, `proj`), where `emb` and `pr` are the
```
```suggestion
the output is a triple (`G`, `emb`, `proj`), where `emb` and `proj` are the
```

simonbrandhorst commented on src/Rings/MPolyQuo.jl:
```diff
@@ -871,7 +871,7 @@ end
 # can be used. However, we are using (graded) quotient rings also with coefficient 
 # rings R which can not be translated to Singular; for instance when R is again 
-# a polynomial ring, or a quotient/localization thereof, or even a `SpecOpenRing`. 
+# a polynomial ring, or a quotient/localization thereof, or even a `AffineSchemeOpenSubschemeRing`. 
```
```suggestion
# a polynomial ring, or a quotient/localization thereof, or even an `AffineSchemeOpenSubschemeRing`. 
```

simonbrandhorst commented on src/deprecations.jl:
```diff
@@ -93,3 +93,40 @@ end
+@deprecate SpecMors AffineSchemeMors
+@deprecate AbsSpecs AbsAffineSchemes
+@deprecate _is_open_func_for_schemes_without_specopen _is_open_func_for_schemes_without_affine_scheme_open_subscheme
+@deprecate spec_open_ring_type affine_scheme_open_subscheme_ring_type
```
```suggestion
```

simonbrandhorst commented on src/exports.jl:
```diff
@@ -1,5 +1,5 @@
 # Entries are sorted with uppercase before lowercase. To resort it,
-# execute:  sort < src/exports.jl > dummy ; mv dummy src/exports.jl
+# execute:  LC_COLLATE=C sort < src/exports.jl > dummy ; mv dummy src/exports.jl
```
Good catch!

simonbrandhorst commented on src/exports.jl:
```diff
@@ -1204,13 +1205,13 @@ export projective_geometry
 export projective_orthogonal_group
 export projective_plane
-export projective_scheme
 export projective_space
```
Please keep the support for `projective_scheme`.
While `proj` is the mathematically correct name for this construction in algebraic geometry.
Oscar is for a wider audience. The name `projective_scheme` is more descriptive of what happens.
(construct a `ProjectiveScheme` from a graded ring). 
Therefore both `proj` and `projective_scheme` should be allowed.
Similarly we should have the constructors `affine_scheme` and `spec`.

lgoettgens replied:
There are already some examples of how to achieve such an alias in Aliases.jl

simonbrandhorst replied:
Maybe someone else wants to use the abbreviations  
`proj` or `spec` in another context with a different meaning. 
Therefore I think that a global alias `proj = projective_scheme` is not justified here?
I merely added calls for `projective_scheme` with the corresponding signatures.

simonbrandhorst committed:
Apply suggestions from code review

simonbrandhorst committed:
add support for projective_scheme and affine_scheme

simonbrandhorst reviewed (approved):
Thanks for the big effort Erik. I appreciate it!

simonbrandhorst committed:
fix typo

simonbrandhorst committed:
fix typo

paemurru commented:
@fingolfin I did not know that @thofma was among OSCAR project leaders, sorry about that. His name does not appear under OSCAR project leaders in the [OSCAR website](https://www.oscar-system.org/contributors/). Perhaps the page should be updated.

> You can of course request that "the OSCAR leaders" sit together and give a final verdict before you change anything, so that you don't end up changing things back and forth (a quite reasonable request).

Yes, this is what I would have liked, as it seemed to me there was no consensus. But considering how everyone is very busy with the OSCAR book and OSCAR 1.0, I didn't think it would be appropriate for me to request this

fingolfin commented:
Yes, the website should be updated. We'll get to it eventually.
